File: lib/qni/export/state_vector_latex.rb
# frozen_string_literal: true

module Qni
  module Export
    # Wraps a symbolic state-vector formula in a standalone LaTeX document.
    class StateVectorLatex
      HEADER_LINES = [
        '\\documentclass[border=2px]{standalone}',
        '',
        '\\usepackage{amsmath}',
        '\\usepackage{amssymb}',
        '\\usepackage{xcolor}',
        '',
        '\\begin{document}'
      ].freeze
      FOOTER_LINES = [
        '\\end{document}'
      ].freeze

      def initialize(latex_formula:, theme: :dark)
        @latex_formula = latex_formula
        @theme = theme
      end

      def render
        (HEADER_LINES + body_lines + FOOTER_LINES).join("\n")
      end

      private

      attr_reader :latex_formula, :theme

      def body_lines
        ["{#{theme_color_line}$\\displaystyle #{latex_formula}$}"]
      end

      def theme_color_line
        "\\color{#{theme_color_name}}"
      end

      def theme_color_name
        case theme
        when :dark
          'white'
        when :light
          'black'
        else
          raise "unsupported export theme: #{theme.inspect}"
        end
      end
    end
  end
end

File: src/export/state_vector_latex.ts
import type { ExportTheme } from './quantikz_latex';

const HEADER_LINES = [
  '\\documentclass[border=2px]{standalone}',
  '',
  '\\usepackage{amsmath}',
  '\\usepackage{amssymb}',
  '\\usepackage{xcolor}',
  '\\newcommand{\\ket}[1]{\\left|#1\\right\\rangle}',
  '',
  '\\begin{document}'
] as const;

const FOOTER_LINES = ['\\end{document}'] as const;

export interface StateVectorLatexOptions {
  readonly latexFormula: string;
  readonly theme: ExportTheme;
}

export class StateVectorLatex {
  private readonly latexFormula: string;
  private readonly theme: ExportTheme;

  constructor(options: StateVectorLatexOptions) {
    this.latexFormula = options.latexFormula;
    this.theme = options.theme;
  }

  render(): string {
    return [...HEADER_LINES, ...this.bodyLines(), ...FOOTER_LINES].join('\n');
  }

  private bodyLines(): string[] {
    return [`{\\color{${this.themeColorName()}}$\\displaystyle ${this.latexFormula}$}`];
  }

  private themeColorName(): string {
    switch (this.theme) {
      case 'dark':
        return 'white';
      case 'light':
        return 'black';
      default:
        throw new Error(`unsupported export theme: ${String(this.theme)}`);
    }
  }
}
